ntoskrnl/config/cmapi.c

```diff
@@ -2251,7 +2251,6 @@ CmUnloadKey(IN PCM_KEY_CONTROL_BLOCK Kcb,
         }
         else
         {
-            DPRINT1("CmUnloadKey: Force unloading is HALF-IMPLEMENTED, expect dangling KCBs problems!\n");
```

Comment:
@GeoB99 Can you check, on real-life cases, what the value of Kcb->RefCount would be _**just after**_ the call to CmpEnumerateOpenSubKeys function, in both cases? (this one and the other above)

Reply:
Force unloading (done in USETUP):
```
WARNING:  HvHiveWillShrink at sdk\lib\cmlib\hivewrt.c:574 is UNIMPLEMENTED!
(ntoskrnl\config\cmapi.c:2264) Kcb->RefCount == 2 (FORCE UNLOAD)
Break instruction exception - code 80000003 (first chance)
nt!CmUnloadKey+0x3bb:
80411b5b cc              int     3
kd> gh
(ntoskrnl\config\cmapi.c:2264) Kcb->RefCount == 3 (FORCE UNLOAD)
Break instruction exception - code 80000003 (first chance)
nt!CmUnloadKey+0x3bb:
80411b5b cc              int     3
kd> gh
(ntoskrnl\config\cmapi.c:2264) Kcb->RefCount == 3 (FORCE UNLOAD)
Break instruction exception - code 80000003 (first chance)
nt!CmUnloadKey+0x3bb:
80411b5b cc              int     3
kd> gh
```
Normal unloading (done during normal startup of the system):
```
(ntoskrnl\config\cmapi.c:2470) Open sub keys: 100
(ntoskrnl\config\cmapi.c:2249) Kcb->RefCount == 5 (FAILED NORMAL UNLOAD)
Break instruction exception - code 80000003 (first chance)
nt!CmUnloadKey+0x25a:
804119fa cc              int     3
kd> gh
(ntoskrnl\config\cminit.c:394) ZwCreateFile(\??\C:\Documents and Settings\NetworkService\ntuser.dat) failed, Status 0xc0000043.
(ntoskrnl\config\cmapi.c:2470) Open sub keys: 92
(ntoskrnl\config\cmapi.c:2249) Kcb->RefCount == 6 (FAILED NORMAL UNLOAD)
Break instruction exception - code 80000003 (first chance)
nt!CmUnloadKey+0x25a:
804119fa cc              int     3
```
Normal unloading (done during log-off):
```
(ntoskrnl\config\cmapi.c:2470) Open sub keys: 27
(ntoskrnl\config\cmapi.c:2249) Kcb->RefCount == 6 (FAILED NORMAL UNLOAD)
Break instruction exception - code 80000003 (first chance)
nt!CmUnloadKey+0x25a:
804119fa cc              int     3
```
The last two cases the unloading process of hive fails because of opened subkeys (all KCBs with RefCount being non-zero, empty cache entries are expunged).

Reply:
Thanks! It's still weird that, in the case of force-unloading, the RefCount of the Kcb (the "root" one corresponding to the hive being unloaded) has still a "high" value, > 1 .